### PR TITLE
Add SPARC and LEON support

### PR DESCRIPTION
**Pull request by ghost**
# Introduction

This patch set adds support for the SPARC architecture and LEON processors.

`arch/sparc` is compliant with the SPARC V8 specification and independent of processor implementation. This includes SPARC System V ABI compliant software run-time in trap handlers and interrupt context. Number of register windows can be configured.

`soc/sparc` provides support for the LEON3 processors implementation, also compatible with LEON4 and LEON5, including the fault-tolerant versions.

More information about the LEON processors and the SPARC specifications can be found at:
<https://www.gaisler.com/index.php/products/processors>

# Boards

## boards/sparc/qemu_leon3

`qemu_leon3` is a board configuration for exercising the LEON3 port with `sanitycheck`.

This board configuration executes in the QEMU simulator distributed with Zephyr SDK-NG.

## boards/sparc/generic_leon3

`generic_leon3` is a board configuration designed to be used with LEON3 FPGA template designs available in the GRLIB/LEON GPL distribution. The GRLIB GPL distribution can be found at: <https://www.gaisler.com/index.php/downloads/leongrlib>

Sanitycheck can run tests for this board using TSIM3.

## boards/sparc/gr716a_mini

The `gr716a_mini` board configuration targets the GR716A fault-tolerant microcontroller.

Sanitycheck runs tests for this board with the TSIM3 LEON simulator.

# Toolchain

The tools provided by the *Zephyr SDK with Crosstool NG* (SDK-NG) SPARC distribution can be used:

* `sdk-ng v0.11.2`
Works out-of-the-box.

* `sdk-ng v0.11.4` and `v0.12.0-beta-1`
Requires the following SDK-NG patch: <https://github.com/zephyrproject-rtos/sdk-ng/pull/288>. (Replace the file `target.cmake` in the SDK installation.)

# Test results
All tests pass:
```
$ ./scripts/sanitycheck --arch sparc
[...]
INFO    - Total complete:  259/ 259  100%  skipped:   67, failed:    0
INFO    - 192 of 192 tests passed (100.00%), 0 failed, 11768 skipped with 0 warnings in 697.39 seconds
````

### CODE REVIEW

**Comment by martin-aberg** [comment]
The Zephyr CI checks now pass, including simulation of the `qemu_leon3` board configuration.

There were initially an issue related to the Zephyr SDK 0.11.3 CMake files which now has a solution.
* Zephyr SDK `v0.11.3`, `v0.11.4`, `v0.12.0-beta-1`
Depends on Zephyr commit 2ac43a6710c5a013989d30f23267dd37c2b8d9af (part of this pull request) to build for SPARC.

* Zephyr SDK after https://github.com/zephyrproject-rtos/sdk-ng/commit/cff96b0bbafe2774df033a1a82049de33518bb4c
Does not depend on 2ac43a6710c5a013989d30f23267dd37c2b8d9af.

This pull request is compatible with the above Zephyr SDK versions.

File: arch/sparc/Kconfig
Comment (andrewboie) [inline arch/sparc/Kconfig:35]: these are some very large stack sizes!

File: arch/sparc/core/interrupt_trap.S
Comment (andrewboie) [inline arch/sparc/core/interrupt_trap.S:153]: swap is the older, more complex context switch method:

```
/**
 * Cooperatively context switch
 *
 * Must be called with interrupts locked with the provided key.
 * This is the older-style context switching method, which is incompatible
 * with SMP. New arch ports, either SMP or UP, are encouraged to implement
 * arch_switch() instead.
 *
 * @param key Interrupt locking key
 * @return If woken from blocking on some kernel object, the result of that
 *         blocking operation.
 */
int arch_swap(unsigned int key);
```

any particular reason why you didn't implement `arch_switch()`?

File: include/arch/sparc/arch.h
Comment (andrewboie) [inline include/arch/sparc/arch.h:102]: I'm assuming the ESF is so small because the interrupted context's register state gets saved through some other method than pushing it onto a stack? (I don't know this arch well)

File: samples/subsys/tracing/sample.yaml
Comment (andrewboie) [inline samples/subsys/tracing/sample.yaml:26]: I'd appreciate filing an Enhancement ticket to eventually get tracing working.

File: CODEOWNERS
Comment (andrewboie) [inline CODEOWNERS:68]: I would add yourself to the MAINTAINERS file as well.

**Review by tyrantlizard77 — COMMENTED** [review]
This all looks really good. Only nitpick really is the use of old-style `arch_swap()` instead of `arch_switch()`. Can you give some detail on why this was chosen? It may become an issue if SMP is later introduced, and someday we'd like to remove the swap support in the core kernel entirely (not happening anytime soon though).

My review does not cover the ASM code, I don't know this CPU well enough to weigh in. Passing tests are good enough for me, at least.

It looks like we already have a SPARC toolchain present in the SDK, which is great.

File: arch/sparc/Kconfig
Comment (martin-aberg) [reply arch/sparc/Kconfig:35]: Yes, stacks are large due to the SPARC ABI stack frames. The stack Kconfig options have been revisited and most of them are reduced in the 2020-11-12 force push.

For example:
```
 config ISR_STACK_SIZE
-       default 8192
+       default 4096

 config TEST_EXTRA_STACKSIZE
-       default 8192 if COVERAGE_GCOV
-       default 4096
+       default 4096 if COVERAGE_GCOV
+       default 2048

 config SYSTEM_WORKQUEUE_STACK_SIZE
-       default 8192
+       default 4096
```

File: arch/sparc/core/interrupt_trap.S
Comment (martin-aberg) [reply arch/sparc/core/interrupt_trap.S]: Thank you for the comments regarding `arch_swap()` and `arch_switch()`. I have updated the implementation so it now uses the the `arch_switch()` API instead.

File: include/arch/sparc/arch.h
Comment (martin-aberg) [reply include/arch/sparc/arch.h:102]: The `struct __esf` type is used only for fatal exceptions in the Zephyr SPARC port. That is, those exceptions which end up in `z_fatal_error()`. So only a minimum of state is saved to the stack because we do not go back from there.

For interrupts on the other hand, more context is saved as required on the stack in `interrupt_trap.S`. The offsets defined in `stack_offsets.h` are used in that case.

File: samples/subsys/tracing/sample.yaml
Comment (martin-aberg) [reply samples/subsys/tracing/sample.yaml]: Agreed about filing enhancement ticket to add tracing. Would it be an appropriate way forward to file that ticket after (if) this pull request is accepted?

The `arch_exclude` in `samples/subsys/tracing/sample.yaml` was specific to the `tracing.osawareness.openocd` sample. It was possible to eliminated the `arch_exclude` by providing the corresponding support in `subsys/debug/openocd.c` instead. That update is also covered by the 2020-11-12 force-push.

File: CODEOWNERS
Comment (martin-aberg) [reply CODEOWNERS:68]: Sure. I will have a look at this. What would the appropriate role be,  maintainer or collaborator?

File: include/arch/sparc/arch.h
Comment (andrewboie) [reply include/arch/sparc/arch.h:102]: If you later add support for CONFIG_USERSPACE, some types of memory protection faults are recoverable. Can cross that bridge when you get there I suppose.

File: samples/subsys/tracing/sample.yaml
Comment (andrewboie) [reply samples/subsys/tracing/sample.yaml]: just open an Enhancement on GitHub summarizing the situation.

File: CODEOWNERS
Comment (andrewboie) [reply CODEOWNERS:68]: If you're willing to be the maintainer, that would be ideal. We can get you added to the GH contributors list so you can be added to code review. cc @MaureenHelm

File: arch/sparc/core/interrupt_trap.S
Comment (andrewboie) [reply arch/sparc/core/interrupt_trap.S]: tyvm!

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by nashif** [comment]
LGTM, can you please some notes in 

doc/releases/release-notes-2.5.rst

So we do not have to chase after you when the release time comes.

File: CODEOWNERS
Comment (MaureenHelm) [reply CODEOWNERS:68]: We need a maintainer, but please start as a collaborator first. You can find more details in the [Project Roles](https://docs.zephyrproject.org/latest/contribute/project_roles.html) document.

File: samples/subsys/tracing/sample.yaml
Comment (martin-aberg) [reply samples/subsys/tracing/sample.yaml]: OK!

File: CODEOWNERS
Comment (martin-aberg) [reply CODEOWNERS:68]: I have now added myself as collaborator in `MAINTAINERS.yml`: b4753af07bc54ad7171be8e740beea210b66f345

**Comment by martin-aberg** [comment]
> LGTM, can you please some notes in
> 
> doc/releases/release-notes-2.5.rst

Yes. The notes are added in 1e4eecde8ffe675923585e505aa7ad7615344b5c.

**Review by nashif — APPROVED** [review]

**Comment by martin-aberg** [comment]
A couple of string formatting tests have started to fail after #29876 was merged earlier today.

* All tests pass when the SPARC patches are applied on cb1cf54406abafb4a781660161a93d5c363c7a1f.

* The following tests fail when SPARC patches are applied on bb99422c8a5f97d250e3e7c956c57ec78fae0e77:
`tests/lib/sprintf/libraries.libc.sprintf`
`tests/kernel/common/kernel.common.misra`
`tests/kernel/common/kernel.common.tls`
`tests/kernel/common/kernel.common`

I am investigating this right now.

File: arch/Kconfig
Comment (carlescufi) [inline arch/Kconfig:45]: I believe this is our first big endian arch. Are all SPARC ICs big endian, this is not configurable like with Arm?

File: boards/sparc/gr716a_mini/doc/index.rst
Comment (carlescufi) [inline boards/sparc/gr716a_mini/doc/index.rst:47]:
```suggestion
The application is linked to the on-chip tightly coupled memory by
```

**Review by carlescufi — APPROVED** [review]
A question and a typo, otherwise this looks great. Thanks!

File: arch/Kconfig
Comment (martin-aberg) [reply arch/Kconfig:45]: Yes, the 32-bit SPARC V8 architecture is always big endian.
(There is also a 64-bit SPARC architecture specification (V9), often referred to as sparc64, which is bi-endian.)

File: boards/sparc/gr716a_mini/doc/index.rst
Comment (martin-aberg) [reply boards/sparc/gr716a_mini/doc/index.rst:47]: Thank you. I will correct the typo as suggested.

**Review by MaureenHelm — APPROVED** [review]

**Comment by martin-aberg** [comment]
> A couple of string formatting tests have started to fail after #29876 was merged earlier today.

This is now fixed in 646d78117104ac1fa5b5ee9b9276224c6bd443c1.

I kept the patch (a one-line removal) in this pull request because the issue was
not reproducible on other platforms before the patch was applied, and the other
platforms still pass all CI tests after the patch is applied.

**Comment by andrewboie** [comment]
This doc failure is weird:
```
scripts/gen_kconfig_rest.py: Kconfig.zephyr:23: '/home/runner/work/zephyr/zephyr/doc/_build/Kconfig/Kconfig.shield.defconfig' not found (in 'source "$(KCONFIG_BINARY_DIR)/Kconfig.shield.defconfig"'). Check that environment variables are set correctly (e.g. $srctree, which is set to '/home/runner/work/zephyr/zephyr'). Also note that unset environment variables expand to the empty string.
```
Unrelated to this PR?

**Comment by martin-aberg** [comment]
> This doc failure is weird:
> 
> ```
> scripts/gen_kconfig_rest.py: Kconfig.zephyr:23: '/home/runner/work/zephyr/zephyr/doc/_build/Kconfig/Kconfig.shield.defconfig' not found (in 'source "$(KCONFIG_BINARY_DIR)/Kconfig.shield.defconfig"'). Check that environment variables are set correctly (e.g. $srctree, which is set to '/home/runner/work/zephyr/zephyr'). Also note that unset environment variables expand to the empty string.
> ```
> 
> Unrelated to this PR?

Yes it is most likely unrelated to this PR. The only doc change in the latest force push was a typo fix inside a `.rst` file (which was not mentioned in the log message).

There was a time window earlier today in which several PRs failed the _Documentation GitHub Workflow_ CI check in the same way. My understanding is that it was fixed in #30002.

**Comment by andrewboie** [comment]
Local doc build with this PR rebased builds fine locally for me.

**Comment by pabigot** [comment]
> > A couple of string formatting tests have started to fail after #29876 was merged earlier today.
> 
> This is now fixed in [646d781](https://github.com/zephyrproject-rtos/zephyr/commit/646d78117104ac1fa5b5ee9b9276224c6bd443c1).
> 
> I kept the patch (a one-line removal) in this pull request because the issue was
> not reproducible on other platforms before the patch was applied, and the other
> platforms still pass all CI tests after the patch is applied.

This patch is correct; thanks.  This would have been a residue after late refactoring of pointer handling.

**Pull request merged**
